### PR TITLE
Bugfixes and test for rehearse

### LINKED ISSUE #7161: nlp.rehearse error in spacy version 3.0.1
I have build custom NER , i just want to add new pattern/ner label in my existing model without forgetting
previous knowledge(incremental/online learning), so that i used rehearse and resume_training method from https://spacy.io/api/language#resume_training

### Environment
* Operating System: Windows
* spaCy Version Used: 3.0.1

### Training Code
```
TRAIN_DATA = [
    ('Who is Kofi Annan?',{
    'entities':[(7,18,'PERSON')]
    }),
    ('Who is Steve Jobs?',{
    'entities':[(7,17,'PERSON')]
    }),
    ('I like London and Berlin.',{
    'entities':[(7,13,'LOC'),(18,24,'LOC')]
    })
]

nlp = spacy.blank('en')

ner = nlp.create_pipe('ner')
nlp.add_pipe('ner')

for _,annotations in TRAIN_DATA:
    for ent in annotations.get('entities'):
        ner.add_label(ent[2])

optimizer = nlp.begin_training()
for itn in range(100):
    for raw_text,entity_offsets in TRAIN_DATA:
        doc = nlp.make_doc(raw_text)
        example = Example.from_dict(doc,entity_offsets)
        nlp.update([example],sgd=optimizer)

nlp.to_disk(output_dir)
```
### Incremental Training Code
I have used rehearse and resume_training method in place of  update method for incremental training (completely new data).
```
training_data = [
    ('Hi Anil',{'entities':[(3,7,'PERSON')]
    }),
    ('Hi Anil, how are you?',{'entities':[(3,7,'PERSON')]
    }),
    ('Hi Ravish, how you doing?',{'entities':[(3,9,'PERSON')]
    })
]

nlp_incremental_try = spacy.load(output_dir)
ner = nlp_incremental_try.get_pipe('ner')

for _,annotations in training_data:
    for ent in annotations.get('entities'):
        ner.add_label(ent[2])

optimizer = nlp_incremental_try.resume_training()
for itn in range(100):
    random.shuffle(training_data)
    for raw_text,entity_offsets in training_data:
        doc = nlp_incremental_try.make_doc(raw_text)
        example = Example.from_dict(doc,entity_offsets)
        nlp_incremental_try.rehearse([example],sgd=optimizer)
```

### Error

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-5-cab349d2a8fa> in <module>
     12         doc = nlp_incremental_try.make_doc(raw_text)
     13         example = Example.from_dict(doc,entity_offsets)
---> 14         nlp_incremental_try.rehearse([example],sgd=optimizer)

~\AppData\Local\Programs\Python\Python38\lib\site-packages\spacy\language.py in rehearse(self, examples, sgd, losses, component_cfg, exclude)
   1170                 continue
   1171             grads = {}
-> 1172             proc.rehearse(
   1173                 examples, sgd=get_grads, losses=losses, **component_cfg.get(name, {})
   1174             )

~\AppData\Local\Programs\Python\Python38\lib\site-packages\spacy\pipeline\transition_parser.pyx in spacy.pipeline.transition_parser.Parser.rehearse()

~\AppData\Local\Programs\Python\Python38\lib\site-packages\spacy\pipeline\trainable_pipe.pyx in spacy.pipeline.trainable_pipe.TrainablePipe.finish_update()

~\AppData\Local\Programs\Python\Python38\lib\site-packages\thinc\model.py in finish_update(self, optimizer)
    322             for name in node.param_names:
    323                 if node.has_grad(name):
--> 324                     param, grad = optimizer(
    325                         (node.id, name), node.get_param(name), node.get_grad(name)
    326                     )

~\AppData\Local\Programs\Python\Python38\lib\site-packages\spacy\language.py in get_grads(W, dW, key)
   1161 
   1162         def get_grads(W, dW, key=None):
-> 1163             grads[key] = (W, dW)
   1164 
   1165         get_grads.learn_rate = sgd.learn_rate

TypeError: unhashable type: 'numpy.ndarray'
```
So kindly suggest me what am i doing wrong?
please help me.

### PR DESCRIPTION
**Pull request by kadarakos**
fixes #7161

### CODE REVIEW

**Review by svlandeg — COMMENTED** [review]
Thanks for the bug fix Ákos!

The `rehearse` functionality has been experimental for some time now, and it would be nice to get this into a better shape. This clear bug fix is certainly a good start.

Could you use the provided example code from issue 7161 to write a test that originally failed, and now succeeds with this bug fix?

**Comment by kadarakos** [comment]
> Could you use the provided example code from issue 7161 to write a test that originally failed, and now succeeds with this bug fix?

That's a very good idea, of course!

**Comment by kadarakos** [comment]
> Could you use the provided example code from issue 7161 to write a test that originally failed, and now succeeds with this bug fix?

The test uses the example from #7161 and now it triggers the `rehearse` for `ner`, `tagger`, `parser` and `textcat_multilabel`. As a result there are also bugfixes included for `tagger` and `textcat_multilabel` that make them **run**, but its not a fix in the sense that it makes them **useful** yet.

File: spacy/tests/training/test_rehearse.py [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            if rehearse:
+                nlp.rehearse([example], sgd=optimizer)
+            else:
+                nlp.update([example], sgd=optimizer)
+    return nlp
+
+
+@pytest.mark.parametrize("component", ['ner', 'tagger', 'parser', 'textcat_multilabel'])
+def test_rehearse(component):
+    print(component)
```
Comment (svlandeg) [inline spacy/tests/training/test_rehearse.py:165]:
```suggestion
```

File: spacy/tests/training/test_rehearse.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+        {
+            'entities': [(3, 9, 'PERSON')],
+            'tags': ['INTJ', 'PROPN', 'PUNCT', 'ADV', 'AUX', 'PRON', 'PUNCT'],
+            'deps': ['intj', 'ROOT', 'punct', 'advmod', 'nsubj', 'advcl', 'punct'],
+            'heads': [1, 1, 1, 5, 5, 1, 1],
+            'morphs': ['', 'VerbForm=Inf', 'PunctType=Comm', '', 'Case=Nom|Person=2|PronType=Prs', 'Aspect=Prog|Tense=Pres|VerbForm=Part', 'PunctType=Peri'],
+            'cats': {'greeting': 1.0, 'question': 1.0}
+        }
+    ),
+    # UTENCIL new label
```
Comment (svlandeg) [inline spacy/tests/training/test_rehearse.py:89]:
```suggestion
    # UTENSIL new label
```

**Review by svlandeg — APPROVED** [review]
Looks good to me, thanks for adding the tests! 🙏

**Comment by honnibal** [comment]
Looks good, and nice test!

**Pull request merged**